tensorflow/core/kernels/generate_box_proposals_op.cu.cc

```diff
@@ -0,0 +1,631 @@
+      y1 = fmax(fmin(y1, img_height - 1.0f), 0.0f);
+      x2 = fmax(fmin(x2, img_width - 1.0f), 0.0f);
+      y2 = fmax(fmin(y2, img_height - 1.0f), 0.0f);
+
+      // Filter boxes
+      // Removing boxes with one dim < min_size
+      // (center of box is in image, because of previous step)
+      width = x2 - x1 + 1.0f;  // may have changed
+      height = y2 - y1 + 1.0f;
+      bool keep_box = fmin(width, height) >= min_size_scaled;
+
+      // We are not deleting the box right now even if !keep_box
+      // we want to keep the relative order of the elements stable
+      // we'll do it in such a way later
+      // d_boxes_keep_flags size: (num_images,prenms_nboxes)
+      // d_out_boxes size: (num_images,prenms_nboxes)
+      const int out_index = image_index * prenms_nboxes + ibox;
+
+      d_boxes_keep_flags[out_index] = keep_box;
+      d_out_boxes[out_index] = {x1, y1, x2, y2};
+      // if(keep_box)printf("Has keep box %d\n",image_index);
+      // d_inout_scores size: (num_images,KA)
+      if (!keep_box)
+        d_inout_scores[image_index * KA + ibox] = FLT_MIN;  // for NMS
+    }
+  }
+}
+
+// Copy the selected boxes and scores to output tensors.
+//
+__global__ void WriteUprightBoxesOutput(
+    const CudaLaunchConfig nboxes, const float4* d_image_boxes,
+    const float* d_image_scores, const int* d_image_boxes_keep_list,
+    const int n_rois, float* d_image_out_rois, float* d_image_out_rois_probs) {
+  CUDA_1D_KERNEL_LOOP(i, nboxes.virtual_thread_count) {
+    if (i < n_rois) {  // copy rois to output
+      const int ibox = d_image_boxes_keep_list[i];
+      const float4 box = d_image_boxes[ibox];
+      const float score = d_image_scores[ibox];
+      // Scattered memory accesses
+      // postnms_nboxes is small anyway
+      d_image_out_rois_probs[i] = score;
+      const int base_idx = 4 * i;
+      d_image_out_rois[base_idx + 0] = box.y;
+      d_image_out_rois[base_idx + 1] = box.x;
+      d_image_out_rois[base_idx + 2] = box.w;
+      d_image_out_rois[base_idx + 3] = box.z;
+    } else {  // set trailing entries to 0
+      d_image_out_rois_probs[i] = 0.;
+      const int base_idx = 4 * i;
+      d_image_out_rois[base_idx + 0] = 0.;
+      d_image_out_rois[base_idx + 1] = 0.;
+      d_image_out_rois[base_idx + 2] = 0.;
+      d_image_out_rois[base_idx + 3] = 0.;
+    }
+  }
+}
+
+// Allocate scratch spaces that are needed for operation
+//
+
+Status AllocateGenerationTempTensors(
+    OpKernelContext* context, Tensor* d_conv_layer_indexes,
+    Tensor* d_image_offset, Tensor* d_cub_sort_buffer,
+    Tensor* d_cub_select_buffer, Tensor* d_sorted_conv_layer_indexes,
+    Tensor* d_sorted_scores, Tensor* dev_boxes, Tensor* dev_boxes_keep_flags,
+    int num_images, int conv_layer_nboxes, size_t cub_sort_temp_storage_bytes,
+    size_t cub_select_temp_storage_bytes, int nboxes_to_generate, int box_dim) {
+  auto d = context->eigen_gpu_device();
+  TF_RETURN_IF_ERROR(context->allocate_temp(
```

Comment:
The pattern:
```
context->allocate_temp(...)
SetZero<<<...>>>(...)
```
occurs >10 times, can we add a helper method for that?

Reply:
Done